### PR TITLE
D8CORE-562: Primary and Secondary Menus

### PR DESCRIPTION
**Pull request by sherakama**
# READY FOR REVIEW

# Summary
- Primary and Secondary Menus
- This is porting the Decanter work in to Stanford Basic while https://github.com/SU-SWS/decanter/pull/534 is being worked out.

# Review By (Date)
- ASAP

# Urgency
- High, Kerri wants to start testing next week.

# Steps to Test

1. TBD

# Affected Projects or Products
- D8CORE

# Associated Issues and/or People
- D8CORE-562
- https://github.com/SU-SWS/decanter/pull/534

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

### CODE REVIEW

**Comment by pookmish** [comment]
I've got 99 file changes and....

**Comment by pookmish** [comment]
is it supposed to show child menu items on hover AND when you click the arrow to the right?

**Comment by sherakama** [comment]
> is it supposed to show child menu items on hover AND when you click the arrow to the right?

Yeah, it's new and needs a bit of iteration. See: https://github.com/SU-SWS/decanter/pull/534#issuecomment-551136316

File: stanford_basic.theme [hunk 1]
```diff
@@ -215,3 +215,36 @@ function stanford_basic_preprocess_page(&$vars) {
 function stanford_basic_preprocess_block__system_branding_block(&$vars) {
   $vars['lockup'] = theme_get_setting('lockup');
 }
+
+/**
+* Implements hook_preprocess_menu() for menu.html.twig.
+*/
+function stanford_basic_preprocess_menu(&$variables, $hook) {
+  if ($hook == 'menu') {
```
Comment (pookmish) [inline stanford_basic.theme]: by using `hook_preprocess_menu` the `$hook`  is always going to be `menu`. dont even need to accept that argument.

File: stanford_basic.theme [hunk 2]
```diff
@@ -215,3 +215,36 @@ function stanford_basic_preprocess_page(&$vars) {
 function stanford_basic_preprocess_block__system_branding_block(&$vars) {
   $vars['lockup'] = theme_get_setting('lockup');
 }
+
+/**
+* Implements hook_preprocess_menu() for menu.html.twig.
```
Comment (pookmish) [inline stanford_basic.theme]: Incorrect docblock formatting. Should be only `Implements hook_preprocess_menu().`

File: stanford_basic.theme [hunk 3]
```diff
@@ -215,3 +215,36 @@ function stanford_basic_preprocess_page(&$vars) {
...
+*/
+function stanford_basic_preprocess_menu(&$variables, $hook) {
+  if ($hook == 'menu') {
+    $current_path = \Drupal::request()->getRequestUri();
+    foreach ($variables['items'] as &$item) {
```
Comment (pookmish) [inline stanford_basic.theme]: This modifieds the admin toolbar menu too. Should check for `menu_name`

File: stanford_basic.theme [hunk 4]
```diff
@@ -215,3 +215,36 @@ function stanford_basic_preprocess_page(&$vars) {
...
+      if ($item['in_active_trail']) {
+        if ($item['url']->toString() == $current_path) {
+          $item['is_active'] = TRUE;
+        } elseif (count($item['below'])) {
```
Comment (pookmish) [inline stanford_basic.theme]: Should be on a new line Implements hook_preprocess_menu()

File: stanford_basic.theme [hunk 5]
```diff
@@ -215,3 +215,36 @@ function stanford_basic_preprocess_page(&$vars) {
...
+
+/**
+ * Set active and active-trail class for sub-menus recursively.
+ */
+function _stanford_basic_menu_process_submenu(&$submenu, $current_path) {
+  foreach ($submenu as &$item) {
+    if ($item['in_active_trail']) {
+      if ($item['url']->toString() == $current_path) {
+        $item['is_active'] = TRUE;
+      } elseif (count($item['below'])) {
```
Comment (pookmish) [inline stanford_basic.theme]: New line Implements hook_preprocess_menu()

File: stanford_basic.theme [hunk 6]
```diff
@@ -215,3 +215,36 @@ function stanford_basic_preprocess_page(&$vars) {
...
+        _stanford_basic_menu_process_submenu($item['below'], $current_path);
+      }
+    }
```
Comment (pookmish) [inline stanford_basic.theme]: isnt this doing the same as stanford_basic_preprocess_menu(). Should consolidate the duplicated code

**Comment by pookmish** [comment]
I'm not going to hold this up on existing issues in codeclimate. but the new ones should be addressed. I think i commented on them

File: stanford_basic.theme [hunk 7]
```diff
@@ -215,3 +215,36 @@ function stanford_basic_preprocess_page(&$vars) {
...
+        }
+      }
+    }
+  }
+}
+
+/**
+ * Set active and active-trail class for sub-menus recursively.
+ */
+function _stanford_basic_menu_process_submenu(&$submenu, $current_path) {
```
Comment (pookmish) [inline stanford_basic.theme]: add param comments in the docblock

**Comment by sherakama** [comment]
@pookmish Thanks for wiring that up. I've got some feedback to work in from Kerri and I will repost.

**Comment by sherakama** [comment]
@pookmish 

Please review it again. Also, please ignore the codeclimate sass/js issues as the will be resolved when the Decanter work gets merged in and I pull all of it back out.

Working site: https://navigation-test.sites.stanford.edu/

Thanks,

File: theme-settings.php [hunk 8]
```diff
@@ -50,7 +42,7 @@ function stanford_basic_form_system_theme_settings_alter(array &$form, FormState
   $form['options_settings']['stanford_basic_lockup'] = [
     '#type' => 'fieldset',
     '#title' => t('Lockup Settings'),
-    '#field_prefix' => "<p>$image_markup</p><p>More examples can be found at: $decanter</p>"
+    '#field_prefix' => "<p>$image_markup</p><p>More examples can be found at: $decanter</p>",
```
Comment (pookmish) [inline theme-settings.php]: Translate the text

File: stanford_basic.theme [hunk 9]
```diff
@@ -215,3 +215,42 @@ function stanford_basic_preprocess_page(&$vars) {
...
+    $menu_item = [$item];
+    _stanford_basic_menu_process_submenu($menu_item, $current_path);
+    $item = $menu_item[0];
+  }
+}
+
+/**
+ * Set active and active-trail class for sub-menus recursively.
+ *
+ * @param  array $submenu      The `$item['below']` structure from a menu array.
```
Comment (pookmish) [inline stanford_basic.theme]: param comments should be an its own line

File: theme-settings.php (hunk 8 above)
Comment (pookmish) [reply theme-settings.php]: also `$image_markup` could be set up in a better method than a regular html string. a render array would be much better. but outside of this PR

**Pull request merged**